### DISCUSSION TITLE
How to play mp3 audio stream from BLE via arduino-audio-tools?

### DISCUSSION THREAD

**Discussion by taxze6** [pschatzmann/arduino-audio-tools#1613] [Q&A] [answered]
### Problem Description

Thank you for the wonderful library!

I am using arduino-audio-tools to try to play mp3 audio streams from BLE.

I am a novice, I have tried to write a lot of program code according to the case, but can not play BLE transmitted mp3 audio stream.

Can you give me some help, thank you 🙏

### Device Description

ESP32 S3 + MAX98375

![751720255064_ pic_hd](https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/00948c1e-87ed-4198-ba0e-d3e332a185a0)

### Sketch

```cpp
#include <SPI.h>
#include <TFT_eSPI.h> 
#include <Wire.h>
#include <BLEDevice.h>
#include <BLEUtils.h>
#include <BLEServer.h>
#include <BMP280.h>
#include <cmath>
#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioBoardStream.h"

using namespace std;
uint8_t* audio;

BLEScan* pBLEScan;
BLECharacteristic* audioCharacteristic;
BLEServer* pServer;
bool deviceConnected = false;      
bool oldDeviceConnected = false;  
#define SERVICE_UUID        "0000A000-0000-1000-8000-00805F9B34FB"
#define writeAudioData_UUID "0000A001-0000-1000-8000-00805F9B34FB"

AudioInfo info(44100, 2, 16);
I2SStream in; 
AudioBoardStream i2s(AudioKitEs8388V1); 
EncodedAudioStream out(&i2s, new MP3DecoderHelix());
StreamCopy copier(out, in); 
Task task("mp3-copy", 10000, 1, 0);

class MyServerCallbacks : public BLEServerCallbacks
{
    void onConnect(BLEServer *pServer)
    {
        deviceConnected = true;
    };

    void onDisconnect(BLEServer *pServer)
    {
        deviceConnected = false;
    }
};

class MyCallbacks: public BLECharacteristicCallbacks {
    void onWrite(BLECharacteristic *pCharacteristic) {
      uint8_t* audio = audioCharacteristic->getData();
      for (int i = 0; i < audio.size(); i++) {
        Serial.print(audio[i]);
      }
      size_t bytes_read = in.readBytes(audio, audio.size());
    }
};

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  auto cfgRx = in.defaultConfig(RX_MODE);
  cfgRx.copyFrom(info);
  in.begin(cfgRx);

  auto config = i2s.defaultConfig(TX_MODE);
  i2s.begin(config);


  out.begin();
  task.begin([](){copier.copy();});


  BLEDevice::init("Robot");
  pServer = BLEDevice::createServer();
  pServer->setCallbacks(new MyServerCallbacks());
  
  BLEService *pService = pServer->createService(SERVICE_UUID);

  audioCharacteristic = pService->createCharacteristic(
                                        writeAudioData_UUID,
                                        BLECharacteristic::PROPERTY_READ |
                                        BLECharacteristic::PROPERTY_WRITE
                                      );                                      

  audioCharacteristic->setCallbacks(new MyCallbacks());
  pService->start();
  BLEAdvertising *pAdvertising = pServer->getAdvertising();
  pAdvertising->start();
  Serial.print("Device ID:");
}
                               
void loop() {
  if (deviceConnected){
    if (!deviceConnected && oldDeviceConnected)
    {
        delay(500);                  
        pServer->startAdvertising(); 
        oldDeviceConnected = deviceConnected;
    }

    if (deviceConnected && !oldDeviceConnected)
    {
        oldDeviceConnected = deviceConnected;
    }
}
```

### Other Steps to Reproduce

How to play unit_8* mp3 data?🙋

### What is your development environment

Arduino

### I have checked existing issues, discussions and online documentation

- [X] I confirm I have checked existing issues, discussions and online documentation

**Comment by pschatzmann** [comment]
This is certenly not a bug so please do not file any issue but use a discussion in the future.

**Reply by pschatzmann** [reply, marked as answer]
I suggest to have a look at [the examples](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/sandbox/ble)

**Reply by taxze6** [reply]
Thank you. I have understood the use of discussion.

I changed my code，Then successfully print out in the serial port:

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/e1f60813-b328-46a3-8341-83fe7f1b63e1)

But for some reason, the audio still isn't playing.

This is my main code right now：

> AudioInfo info(44100, 2, 16);
I2SStream in; 
I2SStream i2s;
//AudioBoardStream i2s(AudioKitEs8388V1); // final output of decoded stream
EncodedAudioStream out(&i2s, new MP3DecoderHelix()); // Decoding stream
StreamCopy copier(out, in); // copy url to decoder

> void setup() {
  // put your setup code here, to run once:
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  auto cfgRx = in.defaultConfig(RX_MODE);
  cfgRx.copyFrom(info);
  in.begin(cfgRx);
  auto cfgTx = i2s.defaultConfig(TX_MODE);
  cfgTx.port_no = 1;
  cfgTx.copyFrom(info);
  cfgTx.i2s_format = I2S_STD_FORMAT;
  cfgTx.pin_bck = 11;
  cfgTx.pin_ws = 10;
  cfgTx.pin_data = 12;
  i2s.begin(cfgTx);
  // setup I2S based on sampling rate provided by decoder
  out.begin();
}

> void loop() {
  copier.copy();
  if (deviceConnected){
    if(audio != NULL){
      for (int i = 0; i < 22464; i++) {
        Serial.print(audio[i]);
      }
      Serial.println();
      size_t bytes_read = in.readBytes(audio, 22464);
      audio = NULL;
    }
}

'audio' is the data I got from Bluetooth BLE.

Can you give me some help? Have a nice day.😊

**Comment by happycth** [comment]
I have encountered a similar problem. Have you made any progress?

**Comment by taxze6** [comment]
I printed these out in the serial port, and it seems that the mp3 data has been decoded successfully and passed into the i2s, but I still have no sound.😭

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/8a19d507-88bf-4c50-9267-16a62164cd14)

**Reply by pschatzmann** [reply]
Are you sure that you did not forget to call begin on the I2S output class ? 

I recommend that you start with a simple Sine generator output sketch, so that you can verify that you understood the processing and that the wiring is correct

**Reply by taxze6** [reply]
Thanks in advance for the assistance.

I'm sure I didn't forget to call the I2S output class.

Then I tried this case: [https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-memory_mp3_short-i2s/streams-memory_mp3_short-i2s.ino](url)

I modified it, and I successfully played the mp3 file.So I'm sure there's nothing wrong with my wiring.

> void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  
  // begin processing
  auto cfgTx = i2s.defaultConfig(TX_MODE);
  cfgTx.sample_rate = 24000;
  cfgTx.channels = 1;
  cfgTx.i2s_format = I2S_STD_FORMAT;
  cfgTx.pin_bck = 11;
  cfgTx.pin_ws = 10;
  cfgTx.pin_data = 12;
  i2s.begin(cfgTx);
  out.begin();
}

The difference between this example and mine is the audio input.Since my audio data is obtained from Bluetooth, I store the data in unit_8*.

The example uses MemoryStream, while I use I2SStream. Therein lies the current difference.

<img width="771" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/c297c08b-5f16-486a-9716-85a07f7d3894">

<img width="751" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/4f986f74-dd6a-4435-bce9-1ba87ece21b9">


Now I don't know what I should modify, can you give me some suggestions, thank you.🙏

**Reply by pschatzmann** [reply]
I am sorry, but I don't understand: what you are doing is not matching with what you are saying:  you copy the I2S input to another I2S output (and not from BLE)!

- Delete the unnecessary I2SStream in: this is just confusing.
- Write the received data to the decoder in the callback and remove your mess that you have in the loop!

[Start with a simple tested first step and then extend your logic step by step
](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working)

**Reply by taxze6** [reply]
I'm sorry I didn't make myself clear.

ble receives a binary stream of mp3 files from the phone's app. I store the received mp3 data in audio.

I just want to know how to play the unit8_* data, which is full of mp3 data, similar to [255,3,22,0,78...].

I refer to some of your examples and write the code of the previous reply, but I can't play the audio. Now I am trying to change I2SStream into AudioStream and MemoryStream, but I don't know if my thinking is correct.

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/c01b2056-2ab0-41b4-9d3b-67e2f98cbead)

Thank you for help me.😊

**Reply by pschatzmann** [reply]
No this does not make any sense! 

Just do what I have written above: write the received data to out in the callback. This decodes the received mp3 data and forwards it to I2S. Make sure to handle the received size properly!

**Reply by pschatzmann** [reply]
Just out of curiosity: what App are you using ?

**Reply by taxze6** [reply]
I wrote the App myself. I'm a software engineer and don't know much about hardware.

This is the mp3 data that my app transmits to esp32 through ble (I do hexadecimal conversion and then print it in the console, and the data sent to esp32 is binary).

<img width="972" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/f2242afe-c404-4fbe-9d8e-ac7aa66de4ac">

<img width="337" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/18083d5a-b16e-4feb-97a5-29c0a0596965">

I modified the program with reference to your reply above.

I have tried to transfer data directly to the mp3 decoder, but it still doesn't play, I don't know what's wrong.

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/276253a0-7ea5-4ce9-9589-0bed8d40f036)

What else should I pay attention to?

**Reply by pschatzmann** [reply]
You still did not move the write to the callback and the handling of the data size is wrong: You need to query audioCharacteristic for the available bytes and then just 

     int written = out.write(audio, len);

Double check that written == len to make sure that all bytes could be processed. 

I would expect that this synchronizes the transmission speed with with the audio processing and you avoid that data is sent too quick and is therefore lost.

The loop should be empty or better contain some delay() to prevent that the loop is starving the other processing

The codec can read the decoded data from an encoded source (specified in the constructor) or you can write the encoded data where the decoding result is forwarded to the destination specified in the constructor.  You are using the second case. This is explained in the [Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio)

**Reply by taxze6** [reply]
<img width="627" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/51d7c319-8e0e-4fc6-a74c-a9c68735567c">

Do you mean to revise it in this way? I put it in the Bluetooth callback.

It also clears the code in loop and adds a delay.

I re-executed the code, but I still didn't hear the audio play

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/5e50c524-adc1-4ed4-a5f8-72f5a54f99f0)

**Comment by pschatzmann** [comment]
The size is wrong: you need to get it from the audioCharacteristic. 
Then test  that the received data matches the sent one...

**Reply by taxze6** [reply]
“You need to query audioCharacteristic for the available bytes” ,I don't know how to handle this,The current 22464 is calculated by my app.

Can you give me an example？😊

**Comment by pschatzmann** [comment]
Call [getLength()](https://github.com/espressif/arduino-esp32/blob/master/libraries/BLE/src/BLECharacteristic.h) if you are using the ESP32 implementation. 

If I remember right the standard MTU is only 20 bytes! So I guess, you are writing 20 bytes of mp3 data and the rest is garbage with each call

**Reply by taxze6** [reply]
<img width="766" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/6aa87401-dd6f-40a2-a268-4ea0e528892b">

Yes, you are right, it seems that the problem is the length.

I got the length using getLength and printed it out. The length is 600.

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/daf0c587-8b76-44f7-8b7b-f0e6c2771c68)

So, do I need to transfer this mp3 file in segments? For App ble segment transfer I know how to do, how should the esp32 receive segment transfer data? Do you have a similar example?

**Reply by pschatzmann** [reply]
I am not sure what you mean with segments: just feed the decoder with the same valid mp3 data that you have in the file, the size of the chunks you feed does not matter: so 600 is just fine

**Reply by taxze6** [reply]
<img width="722" alt="image" src="https://github.com/pschatzmann/arduino-audio-tools/assets/71701456/88ce81b2-266c-48fd-a654-f35d26cf5d90">

I have modified the code of the app to transfer the mp3 data in segments, sending 20 bytes each time.

You mean, if the App only sends 20 bytes at a time but keeps happening, the esp32 only needs to keep typing 20 bytes through out.write?

**Reply by taxze6** [reply]
Yes, after this modification, I have made progress, and now the speaker is starting to sound, but it is all noise, is it because 20 bytes is too small to play correctly?

**Reply by taxze6** [reply]
I modified to transfer 500 bytes at a time, and now I can faintly hear the sound playing in the mp3. If I want to play the mp3 audio smoothly, do I need to splice all the files together in the esp32 and then decode it?

**Reply by taxze6** [reply]
I did the subcontracting in the app, the esp32 did the splicing function, and finally decoded, and now I can play the audio normally! Thank you for your help～

**Reply by pschatzmann** [reply]
BLE is very inefficient, so make the transfer size as big as possible and minimize the amount of data to be transferred.

If you can't keep up with the speed, you need to store the data in a queue and start with the playback only after you received all data.

**Reply by pschatzmann** [reply]
I am a bit confused: what do you mean with subcontracting and splicing. I tried to google the topic but did not find anything. Can you provide some reference or explanation ?

**Reply by taxze6** [reply]
I'm sorry that my English is not very good, so you have some misunderstanding.

What this means is that when I transmit audio in my app, I break the audio down into many small chunks, and each small chunk is 100 bytes. These blocks are then accepted in the esp32 and synthesized into audio. The complete audio data is then put into the mp3 decoder.

You can see my previous reply, there is a piece of code about the app to send audio data in segments.😊
